[P2P] Timeout for headers sync

sdaftuar opened the pull request:
This first commit delays requesting blocks from peers unless their tip has more than `nMinimumChainWork` (our anti-DoS threshold).  This speeds up initial headers sync by delaying block downloads from starting until we're closer to being caught up.

The second commit implements a timeout on the initial headers sync: when we send the initial getheaders to a peer, estimate the number of headers we expect to receive, and set a timeout (15 minutes + 1 ms / header).

This is a pretty naive algorithm, but I think the worst case behavior (eg if we're bottlenecked by our own bandwidth) is that we cycle through peers while downloading headers, at most every 15 minutes -- which I think isn't too awful, since those 15 minute windows should be making progress at least, and this bounds the time a bad peer can interfere with headers sync. 

FYI I kicked off a fresh node with this patch, and it took me about 1 minute to have the whole headers chain.

Implements #9068, though this isn't the exact approach @sipa suggested in that issue.

sdaftuar committed:
Delay parallel block download until chain has sufficient work

nMinimumChainWork is an anti-DoS threshold; wait until we have a proposed
tip with more work than that before downloading blocks towards that tip.

gmaxwell commented:
Concept ACK.  I suppose this timeout could later be made dynamic like has been proposed block transfer timeouts.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -3211,6 +3215,24 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
+                if (nNow > state.nHeadersSyncTimeout && !pto->fWhitelisted && nSyncStarted == 1 && (nPreferredDownload - state.fPreferredDownload >= 1)) {
+                    // Disconnect a (non-whitelisted) peer if it is our only sync peer,
+                    // and we have others we could be using instead.
+                    LogPrintf("Timeout downloading headers from peer=%d, disconnecting\n", pto->id);
```
```gcc
error: ‘const NodeId CNode::id’ is private within this context

paveljanik replied:
It should be `pto->GetId()`.

sdaftuar commented:
Rebased after #10189.

sdaftuar committed:
Add timeout for headers sync

At startup, we choose one peer to serve us the headers chain, until
our best header is close to caught up.  Disconnect this peer if more
than 15 minutes + 1ms/expected_header passes and our best header
is still more than 1 day away from current time.

jonasschnelli commented on src/net_processing.cpp:
```diff
@@ -3206,6 +3210,24 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
+                    return true;
+                }
+            } else {
+                // Once we've caught up once, reset the timeout so we can't trigger
```
nit: `Once we've caught up once`

sdaftuar replied:
Improved the comment in 898fc2ac689aa77e74362aa27d12120256c1942e.

jonasschnelli commented:
utACK ca4aebd3c50bc424afd25a5611a6ea1e77429413

For the hybrid SPV mode (#9483), a fast headers sync is crucial and also for normal operations, someone could significant delay your initial sync.
If you connect to a malicious or just a very slow peer (It happened to me a couple of times), couldn't you  – in theory – spend a couple of hours syncing headers? Even after this PR, I guess a peer could serve you the 2k headers package every 10minutes (while pong accordingly) leading to a header sync of more then 30h.

I'm not sure if this is solvable without to much of effort, maybe the only solution would be parallel headers download... or at least a check among a handful of peers which could serve 3x2000 headers the fastest.

sdaftuar commented:
> If you connect to a malicious or just a very slow peer (It happened to me a couple of times), couldn't you – in theory – spend a couple of hours syncing headers? Even after this PR, I guess a peer could serve you the 2k headers package every 10minutes (while pong accordingly) leading to a header sync of more then 30h.

@jonasschnelli After this PR, if you have started syncing headers from your initial sync peer and your pindexBestHeader isn't caught up to within 24 hours of current time, then you'll disconnect them.  So this puts a timeout on the entire headers sync (a bit over 20 minutes, currently -- 15 minutes + 1 ms/[expected number of headers]).

Once our headers chain is within 24 hours of current time, we'll request headers from our other peers as well, so after that point as long as one of your peers isn't stalling you, you'll get the headers chain.

jonasschnelli commented:
Thanks @sdaftuar for the explanation. Make sense.
I thought the timer resets after each headers message. A 20min delay hurts but is probably unavoidable (the fact that we don't know the bandwidth limitations and potentials).

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3206,6 +3210,24 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
+        if (state.fSyncStarted && state.nHeadersSyncTimeout < std::numeric_limits<int64_t>::max()) {
+            // Detect whether this is a stalling initial-headers-sync peer
+            if (pindexBestHeader->GetBlockTime() <= GetAdjustedTime() - 24*60*60) {
+                if (nNow > state.nHeadersSyncTimeout && !pto->fWhitelisted && nSyncStarted == 1 && (nPreferredDownload - state.fPreferredDownload >= 1)) {
```
I believe usually its good practice to print if we would otherwise have disconnected a whitelisted peer.

sdaftuar replied:
Fixed in a27ff02f537a3225372e1e1286f0a93fb9037ba9

TheBlueMatt commented on src/net_processing.h:
```diff
@@ -17,6 +17,10 @@ static const int64_t ORPHAN_TX_EXPIRE_TIME = 20 * 60;
 static const unsigned int DEFAULT_BLOCK_RECONSTRUCTION_EXTRA_TXN = 100;
+/** Headers download timeout expressed in microseconds
+ *  Timeout = base + per_header * (expected number of headers) */
+static constexpr int64_t HEADERS_DOWNLOAD_TIMEOUT_BASE = 15 * 60 * 1000000; // 15 minutes
```
If I can take this opportunity to bikeshed your constants - why even bother with a base timeout this high? maybe just a minute (to handle the really-close-to-done-dont-want-a-1ms-timeout case) and then double/tripple the TIMEOUT_PER_HEADER?

sdaftuar replied:
I just wanted to do something clearly safe (it's easy to reason about booting your sync peer potentially every 15 minutes, but slightly scarier to think there could be a circumstance you'd boot that peer every 1 minute).

Anyway I'd like to see what other opinions are, I'm fine adjusting them however people would like...

TheBlueMatt reviewed (commented):
utACK 898fc2ac689aa77e74362aa27d12120256c1942e, few notes, but none matter much.

sdaftuar commented:
Addressed @TheBlueMatt's nit in a fixup commit, let me know if I can squash.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -3206,6 +3210,28 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
+        if (state.fSyncStarted && state.nHeadersSyncTimeout < std::numeric_limits<int64_t>::max()) {
+            // Detect whether this is a stalling initial-headers-sync peer
+            if (pindexBestHeader->GetBlockTime() <= GetAdjustedTime() - 24*60*60) {
+                if (nNow > state.nHeadersSyncTimeout && nSyncStarted == 1 && (nPreferredDownload - state.fPreferredDownload >= 1)) {
```
`nPreferredDownload - state.fPreferredDownload >= 1` was slightly confusing for me. It's saying we should only timeout a peer if we have another peer which is preferred. This doesn't exactly match the logic in `SendMessages()` where we'll try to start block sync from a non-preferred peer if we don't have any better options.

I think this is fine. It's not obvious that we should time out a peer if we don't have any other options which are preferred, but perhaps a comment here would make this clearer.

sdaftuar replied:
Added a comment in 595d51c

jnewbery reviewed (commented):
utACK a27ff02f537a3225372e1e1286f0a93fb9037ba9 with one request for a comment.

jnewbery commented:
reACK https://github.com/bitcoin/bitcoin/commit/595d51c2b746c4473a02e0417dd42c840b1f589c

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -3206,6 +3210,31 @@ bool SendMessages(CNode* pto, CConnman& connman, const std::atomic<bool>& interr
+                        pto->fDisconnect = true;
+                        return true;
+                    } else {
+                        LogPrintf("Timeout downleading headers from whitelisted peer=%d, not disconnecting\n", pto->GetId());
```
Additionally, it would be nice to unset this peer as fSyncStarted so that we pick another peer to donwnload from, in case the whitelisted sync peer is broken.

sdaftuar replied:
Addressed both of these comments in b4c8b5e

TheBlueMatt reviewed (commented):
Discussed making the whole download logic a much clearer state machine, but that can come after this.

sdaftuar commented:
Squashed 65e300812 -> 76f74811c

laanwj commented:
utACK 76f7481
